File: app/src/main/java/com/syroniko/casseteapp/ChatAndMessages/ChatAdapter.java
package com.syroniko.casseteapp.ChatAndMessages;

import android.content.Context;
import android.content.Intent;
import android.provider.ContactsContract;
import android.text.format.DateFormat;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.syroniko.casseteapp.MainClasses.User;
import com.syroniko.casseteapp.R;

import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public class ChatAdapter extends RecyclerView.Adapter<ChatAdapter.ViewHolder>{
    private Context mContext;
    public static final int MESSAGE_RIGHT=0;
    public static final int MESSAGE_LEFT=1;
    private List<Chat> mChat;
    private String imageUrl;
    FirebaseUser fUser;

    public ChatAdapter (Context mContext,List<Chat> mChat,String ImageUrl){
        this.mChat=mChat;
        this.mContext=mContext;
        this.imageUrl= ImageUrl;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if(viewType==MESSAGE_RIGHT){
            Log.d("SASA","SUSA");
            View view= LayoutInflater.from(mContext).inflate(R.layout.chat_item_right,parent,false);
            return new ChatAdapter.ViewHolder(view);
        }
        else {
            Log.d("Sasa","Sisa");
            View view = LayoutInflater.from(mContext).inflate(R.layout.chat_item_left, parent, false);
            return new ChatAdapter.ViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Chat chat =mChat.get(position);
        holder.show_message.setText(chat.getMessage());
        holder.timestampTv.setVisibility(View.INVISIBLE);
        if(imageUrl=="default"){
         //   holder.userImage.setImageResource(R.mipmap.ic_launcher);
        }
        else{
      //      Glide.with(mContext).load(imageUrl).into(holder.userImage);
        }
        if(position==mChat.size()-1){
            Log.d("cmoon",String.valueOf(chat.isSeen()));
            if(chat.isSeen()){
                holder.messageSeen.setText("Seen");
            }
            else{
                holder.messageSeen.setText("Delivered");


            }
        }
        else{
            holder.messageSeen.setVisibility(View.INVISIBLE);
        }
        String time=chat.getTimestamp();
        Calendar cal=Calendar.getInstance(Locale.ENGLISH);
        cal.setTimeInMillis(Long.parseLong(time));
        String dateTime= DateFormat.format("hh:mm",cal).toString();
        holder.timestampTv.setText(dateTime);

    }

    @Override
    public int getItemCount() {
        return mChat.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public TextView show_message;
        public ImageView userImage;
        public TextView messageSeen;
        public TextView timestampTv;

        public ViewHolder(View itemView){
            super(itemView);
            show_message=itemView.findViewById(R.id.chat_item_textView_bubble);
            userImage=itemView.findViewById(R.id.message_fragment_recycler_item_user_image);
            messageSeen=itemView.findViewById(R.id.message_seen_text);
            timestampTv=itemView.findViewById(R.id.timestamptextviewchat);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            Log.v("gottem","gorrwm");
            if(timestampTv.getVisibility()==View.INVISIBLE) {
                timestampTv.setVisibility(View.VISIBLE);
            }
            else{
                timestampTv.setVisibility(View.INVISIBLE);
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        fUser= FirebaseAuth.getInstance().getCurrentUser();
        if(mChat.get(position).getSender().equals(fUser.getUid())){
            return MESSAGE_RIGHT;
        }
        else{
            return MESSAGE_LEFT;
        }
    }
}

File: app/src/main/java/com/syroniko/casseteapp/chatAndMessages/ChatAdapter.kt

package com.syroniko.casseteapp.chatAndMessages

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.RelativeLayout
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.firebase.ktx.Firebase
import com.google.firebase.storage.ktx.storage
import com.syroniko.casseteapp.mainClasses.toast
import com.syroniko.casseteapp.R
import com.syroniko.casseteapp.chatAndMessages.entities.ImageMessage
import com.syroniko.casseteapp.chatAndMessages.entities.Message
import com.syroniko.casseteapp.chatAndMessages.entities.MessageType
import com.syroniko.casseteapp.chatAndMessages.entities.TextMessage
import com.syroniko.casseteapp.firebase.Auth
import com.syroniko.casseteapp.utils.*
import dagger.hilt.android.qualifiers.ApplicationContext
import java.text.SimpleDateFormat
import java.util.*
import javax.inject.Inject


class ChatAdapter @Inject constructor(
    @ApplicationContext val context: Context
) : RecyclerView.Adapter<ChatViewHolder>() {

    var messages: MutableList<Message> = mutableListOf()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    private val uid = Auth.getUid()
    lateinit var imageUrl: String

    private var maxWidth: Double = 0.0
    private var maxHeight: Double = 0.0

    init {
        val maxWidthAndHeight = getMaxWidthAndHeight(context)
        maxWidth = maxWidthAndHeight[0]
        maxHeight = maxWidthAndHeight[1]
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChatViewHolder {
        val view: View
        val type: MessageType

        when (viewType) {
            MESSAGE_TEXT_RIGHT -> {
                view = LayoutInflater.from(context).inflate(R.layout.chat_item_text_right, parent, false)
                type = MessageType.TEXT
            }
            MESSAGE_TEXT_LEFT -> {
                view = LayoutInflater.from(context).inflate(R.layout.chat_item_text_left, parent, false)
                type = MessageType.TEXT
            }
            MESSAGE_IMAGE_RIGHT -> {
                view = LayoutInflater.from(context).inflate(R.layout.chat_item_image_right, parent, false)
                type = MessageType.IMAGE
            }
            MESSAGE_IMAGE_LEFT -> {
                view = LayoutInflater.from(context).inflate(R.layout.chat_item_image_left, parent, false)
                type = MessageType.IMAGE
            }
            else -> {
                view = LayoutInflater.from(context).inflate(R.layout.chat_item_spotify_track, parent, false)
                type = MessageType.SPOTIFY_TRACK
            }
        }

        return ChatViewHolder(view, type)
    }

    override fun onBindViewHolder(holder: ChatViewHolder, position: Int) {
        if (holder.messageType == MessageType.TEXT) {
            holder.bindTextMessage(context, messages[position] as TextMessage)
        }
        else if (holder.messageType == MessageType.IMAGE){
            holder.bindImageMessage(context, messages[position] as ImageMessage, maxWidth, maxHeight)
        }
        else if (holder.messageType == MessageType.SPOTIFY_TRACK){
            //do nothing
        }

        if (position < messages.size - 1){
            if (messages[position].senderId != messages[position+1].senderId){
                addImage(context, messages[position].senderId, holder.userImage)
                holder.userImage.visibility = View.VISIBLE
            }
            else{
                if (messages[position].senderId != uid){
                    holder.userImage.visibility = View.INVISIBLE
                }
            }
        }
        else{
            addImage(context, messages[position].senderId, holder.userImage)
            holder.userImage.visibility = View.VISIBLE
        }

        if (messages[position].senderId == uid){
            holder.userImage.visibility = View.GONE
        }

    }

    override fun getItemCount() = messages.size

//    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
//        val showMessage: TextView = itemView.findViewById(R.id.chat_item_textView_bubble)
//        val userImage: ImageView = itemView.findViewById(R.id.profile_image_chat_activity)
//        val messageSeen: TextView = itemView.findViewById(R.id.message_seen_text)
//        val timestampTv: TextView = itemView.findViewById(R.id.timestamptextviewchat)
//    }

    override fun getItemViewType(position: Int): Int {
        val uid = Auth.getUid()
        val message = messages[position]
        return if((message is TextMessage) && message.senderId == uid){
            Log.d(ChatAdapter::class.simpleName, "RIGHT")
            MESSAGE_TEXT_RIGHT;
        }
        else if((message is TextMessage) && message.senderId != uid){
            Log.d(ChatAdapter::class.simpleName, uid ?: "LEFT")
            MESSAGE_TEXT_LEFT;
        }
        else if(message is ImageMessage && message.senderId == uid){
            MESSAGE_IMAGE_RIGHT
        }
        else if(message is ImageMessage && message.senderId != uid){
            MESSAGE_IMAGE_LEFT
        }
        else {
            MESSAGE_SPOTIFY_TRACK
        }
    }



    companion object {
        const val MESSAGE_TEXT_RIGHT = 0
        const val MESSAGE_TEXT_LEFT = 1
        const val MESSAGE_IMAGE_RIGHT = 2
        const val MESSAGE_IMAGE_LEFT = 3
        const val MESSAGE_SPOTIFY_TRACK = 4
    }

}